diff --git a/CMakeLists.txt b/CMakeLists.txt
index 06bc77c..9ac12fc 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -12,8 +12,8 @@ option(INSTALL_STATIC_LIBS "Install static libraries" OFF)
 option(WITH_SYSTEM_ZLIB "Use system provided zlib library" OFF)
 
 if(CMAKE_C_COMPILER_ID MATCHES "GNU")
-  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fvisibility=hidden -fPIC -O3 -flto")
-  set(CMAKE_INTERPROCEDURAL_OPTIMIZATION ON)
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fvisibility=hidden -fPIC -O3")
+  set(CMAKE_INTERPROCEDURAL_OPTIMIZATION OFF)
 elseif(CMAKE_C_COMPILER_ID MATCHES "Clang")
   set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -O3")
   set(CMAKE_INTERPROCEDURAL_OPTIMIZATION OFF)
@@ -34,7 +34,11 @@ add_subdirectory(deps/lzma-19.00 EXCLUDE_FROM_ALL)
 
 # zlib
 if (WITH_SYSTEM_ZLIB)
-  pkg_check_modules(ZLIB REQUIRED zlib)
+  if (WINDOWS_STORE)
+    find_package(ZLIB REQUIRED)
+  else()
+    pkg_check_modules(ZLIB REQUIRED zlib)
+  endif()
   list(APPEND PLATFORM_INCLUDES ${ZLIB_INCLUDE_DIRS})
   list(APPEND PLATFORM_LIBS ${ZLIB_LIBRARIES})
 else()
@@ -68,6 +72,10 @@ if (INSTALL_STATIC_LIBS)
   )
 endif()
 
+if(MSVC)
+    target_compile_definitions(chdr-static PRIVATE _CRT_SECURE_NO_WARNINGS)
+endif()
+
 if (BUILD_SHARED_LIBS)
   set(CMAKE_CXX_VISIBILITY_PRESET hidden)
   set(CMAKE_VISIBILITY_INLINES_HIDDEN 1)
@@ -79,7 +87,7 @@ if (BUILD_SHARED_LIBS)
 
   if(MSVC)
     target_compile_definitions(chdr PUBLIC "CHD_DLL")
-    target_compile_definitions(chdr PRIVATE "CHD_DLL_EXPORTS")
+    target_compile_definitions(chdr PRIVATE "CHD_DLL_EXPORTS" _CRT_SECURE_NO_WARNINGS)
   elseif(APPLE)
     target_link_options(chdr PRIVATE -Wl,-dead_strip -Wl,-exported_symbol,_chd_*)
   else()
 
